tarantool_test.go

```diff
@@ -660,6 +660,105 @@ func TestClient(t *testing.T) {
 	}
 }
 
+func TestClientSessionPush(t *testing.T) {
+	var (
+		conn *Connection
+		fut  *Future
+		it   ResponseIterator
+		resp *Response
+		err  error
+	)
+
+	conn, err = Connect(server, opts)
+	if err != nil {
+		t.Errorf("Failed to connect: %s", err.Error())
+		return
+	}
+	if conn == nil {
+		t.Errorf("conn is nil after Connect")
+		return
+	}
+	defer conn.Close()
+
+	// Future.Get ignores push messages.
+	const pushCnt1 = 3
+	fut = conn.Call17Async("push_func", []interface{}{pushCnt1})
+	resp, err = fut.Get()
+	if err != nil {
+		t.Errorf("Failed to Call: %s", err.Error())
+	} else if resp == nil {
+		t.Errorf("Response is nil after CallAsync")
+	} else if len(resp.Data) != 1 {
+		t.Errorf("Response.Data is empty after CallAsync")
+	} else if resp.Data[0].(uint64) != pushCnt1 {
+		t.Errorf("result is not {{1}} : %v", resp.Data)
+	}
+
+	// The code iterates through the ready response.
+	var pushCnt, respCnt uint64
+	for it = fut.GetIterator(); it.Next(); {
+		resp = it.Value()
+		if resp.Code == PushCode {
+			pushCnt += 1
+			if resp.Data[0].(uint64) != pushCnt {
+				t.Errorf("Unexpecetd push data = %v", resp.Data)
+			}
+		} else {
+			respCnt += 1
+			if len(resp.Data) != 1 {
+				t.Errorf("Response.Data is empty after CallAsync")
+			} else if resp.Data[0].(uint64) != pushCnt1 {
+				t.Errorf("result is not {{1}} : %v", resp.Data)
+			}
+		}
+	}
+	if err = it.Err(); err != nil {
+		t.Errorf("An unexpected iteration error: %s", err.Error())
+	}
+	if pushCnt != pushCnt1 {
+		t.Errorf("Expect %d pushes but got %d", pushCnt1, pushCnt)
+	}
+	if respCnt != 1 {
+		t.Errorf("Expect %d responses but got %d", 1, respCnt)
+	}
+
+	// The code iterates through a non-ready response with a timeout.
```

Comment:
And what are we going to test? That count of responses (push messages + a response) >= 0 && <= cnt + 1? I added such test with iteration without timeout && fiber.sleep().
```patch
--- a/tarantool_test.go
+++ b/tarantool_test.go
@@ -632,12 +632,36 @@ func TestClientSessionPush(t *testing.T) {
        conn := connect(t, server, opts)
        defer conn.Close()
 
+       var it ResponseIterator
        const pushMax = 3
+       // It will iterate immediately.
+       fut0 := conn.Call17Async("push_func", []interface{}{pushMax})
+       respCnt := 0
+       for it = fut0.GetIterator(); it.Next(); {
+               err := it.Err()
+               resp := it.Value()
+               if err != nil {
+                       t.Errorf("Unexpected error after it.Next() == true: %q", err.Error())
+                       break
+               }
+               if resp == nil {
+                       t.Errorf("Response is empty after it.Next() == true")
+               }
+               respCnt += 1
+       }
+       if err := it.Err(); err != nil {
+               t.Errorf("An unexpected iteration error: %s", err.Error())
+       }
+       if respCnt < 0 || respCnt > pushMax + 1 {
+               t.Errorf("Unexpected respCnt = %d, expected 0 <= respCnt <= %d", respCnt, pushMax + 1)
+       }
```

Our tests depends on Opts.Timeout value. They can sometimes fail (CI/CD container was paused by a time > Opts.Timeout) and it's ok. In fact in our tests we can 'guarantee' only that count of responses (push messages + a response) will be >= 0, nothing more. If we add more timeouts to our tests, it will only be more difficult to 'guarantee' their work when we expect it.

Reply:
> Is count of responses (push messages + a response) >= 0 && <= cnt + 1?

Isn't `it.Next()` waiting until next push value?

Well, if timeouts are adequate the possibility of tests failing because of some external disturbance is really low (I haven't seen any issues in this repo since we introduced CI/CD). You may increase them here if you're wish.

Green tests are good, but it is even better to have tests that replicate some real case. And I think that "long" request with pushes is closer to reality than consecutive pushing in cycle without any other work. It is highly likely that this "non-ready" request in fact is a "ready" request because Lua procedure is really fast and test may not cover the case we're expecting.

Reply:
No, it doesn't wait if no timeout. See an example from https://www.tarantool.io/en/doc/latest/reference/reference_lua/box_session/push/ for an asynchronous case.

It still won't give guarantees. Opts.Timeout = 1s, 2s or 10m just increase the chance of a successful result.

And yet, it seems to me timeouts and sleeps is very bad idea for unit tests. I suggest testing the iterator in a different way:

Move Future.appendPushes/Future.setResponse/Future.setError into the public API. And write unit-tests for Future + Future.GetResponseIterator() without Connection, Tarantool etc.

Reply:
> Green tests are good, but it is even better to have tests that replicate some real case

Please see future_test.go after the update. Yes, we still don't test connection timeouts (we don't do that anywhere). But the iterator itself is tested for this case. That may be enough.

Much more refactoring will be needed to test connection timeouts (and we need to think about how to do it). It seems to me that it is better to do this in a separate issue.

Reply:
Future unit tests are detailed and should cover all situations including one I've wanted do. Thank you, I think then this issue is resolved.